nomad/structs/services.go

```diff
@@ -0,0 +1,692 @@
+}
+
+// ValidateName checks if the services Name is valid and should be called after
+// the name has been interpolated
+func (s *Service) ValidateName(name string) error {
+	// Ensure the service name is valid per RFC-952 §1
+	// (https://tools.ietf.org/html/rfc952), RFC-1123 §2.1
+	// (https://tools.ietf.org/html/rfc1123), and RFC-2782
+	// (https://tools.ietf.org/html/rfc2782).
+	re := regexp.MustCompile(`^(?i:[a-z0-9]|[a-z0-9][a-z0-9\-]{0,61}[a-z0-9])$`)
+	if !re.MatchString(name) {
+		return fmt.Errorf("service name must be valid per RFC 1123 and can contain only alphanumeric characters or dashes and must be no longer than 63 characters: %q", name)
+	}
+	return nil
+}
+
+// Hash returns a base32 encoded hash of a Service's contents excluding checks
+// as they're hashed independently.
+func (s *Service) Hash(allocID, taskName string, canary bool) string {
+	h := sha1.New()
+	io.WriteString(h, allocID)
+	io.WriteString(h, taskName)
+	io.WriteString(h, s.Name)
+	io.WriteString(h, s.PortLabel)
+	io.WriteString(h, s.AddressMode)
+	for _, tag := range s.Tags {
+		io.WriteString(h, tag)
+	}
+	for _, tag := range s.CanaryTags {
+		io.WriteString(h, tag)
+	}
+
+	// Vary ID on whether or not CanaryTags will be used
+	if canary {
+		h.Write([]byte("Canary"))
+	}
+
+	// Base32 is used for encoding the hash as sha1 hashes can always be
+	// encoded without padding, only 4 bytes larger than base64, and saves
+	// 8 bytes vs hex. Since these hashes are used in Consul URLs it's nice
+	// to have a reasonably compact URL-safe representation.
+	return b32.EncodeToString(h.Sum(nil))
+}
+
+// Equals returns true if the structs are recursively equal.
+func (s *Service) Equals(o *Service) bool {
+	if s == nil || o == nil {
+		return s == o
+	}
+
+	if s.AddressMode != o.AddressMode {
+		return false
+	}
+
+	if !helper.CompareSliceSetString(s.CanaryTags, o.CanaryTags) {
+		return false
+	}
+
+	if len(s.Checks) != len(o.Checks) {
+		return false
+	}
+
+OUTER:
+	for i := range s.Checks {
+		for ii := range o.Checks {
+			if s.Checks[i].Equals(o.Checks[ii]) {
+				// Found match; continue with next check
+				continue OUTER
+			}
+		}
+
+		// No match
+		return false
+	}
+
+	if !s.Connect.Equals(o.Connect) {
+		return false
+	}
+
+	if s.Name != o.Name {
+		return false
+	}
+
+	if s.PortLabel != o.PortLabel {
+		return false
+	}
+
+	if !helper.CompareSliceSetString(s.Tags, o.Tags) {
+		return false
+	}
+
+	return true
+}
+
+// ConsulConnect represents a Consul Connect jobspec stanza.
+type ConsulConnect struct {
+	// Native is true if a service implements Connect directly and does not
+	// need a sidecar.
+	Native bool
+
+	// SidecarService is non-nil if a service requires a sidecar.
+	SidecarService *ConsulSidecarService
+}
+
+// Copy the stanza recursively. Returns nil if nil.
+func (c *ConsulConnect) Copy() *ConsulConnect {
+	if c == nil {
+		return nil
+	}
+
+	return &ConsulConnect{
+		Native:         c.Native,
+		SidecarService: c.SidecarService.Copy(),
```

Comment:
Do we need to have a `Canonicalize` method for `ConsulConnect` to ensure this isn't a nil-pointer as well? Especially given that the `Validate` method below allows it to be `nil`. The test [here](https://github.com/hashicorp/nomad/pull/6045/files#diff-409bd77f24264d026ecaeda7aeadedceR22) hands it a non-nil but empty struct which avoids a nil-pointer.

(This seems to show up in `ConsulProxy` below as well. Not sure if I'm being too belt-and-suspenders here though.)

Reply:
It's a bit unfortunate but ConsulConnect and other connect structs being nil are valid identifiers indicating different features are enabled disabled.  I'll add nil checks here.